internal/controllers/machine/machine_controller.go

```diff
@@ -741,8 +742,19 @@ func (r *Reconciler) reconcileDeleteExternal(ctx context.Context, m *clusterv1.M
 	return obj, nil
 }
 
+// shouldAdopt returns true if the Machine should be adopted as a stand-alone Machines directly owned by the Cluster.
 func (r *Reconciler) shouldAdopt(m *clusterv1.Machine) bool {
-	return metav1.GetControllerOf(m) == nil && !util.HasOwner(m.OwnerReferences, clusterv1.GroupVersion.String(), []string{"Cluster"})
+	// if the machine is not yet controller by something, or if it is not already been adopted as a stand-alone machine directly owned by the Cluster.
+	if metav1.GetControllerOf(m) == nil && !util.HasOwner(m.OwnerReferences, clusterv1.GroupVersion.String(), []string{"Cluster"}) {
+		// If the Machine is originated by a MachineDeployment, prevents it to be adopted as a stand alone Machine.
+		// Note: this is required because after restore from a backup both the Machine controller and the
+		// MachineSet controller are racing to adopt Machines, see https://github.com/kubernetes-sigs/cluster-api/issues/7529
```

Comment:
```suggestion
		// If the Machine is originated by a MachineDeployment, this prevents it from being adopted as a stand alone Machine.
		// Note: this is required because after restore from a backup both the Machine controller and the
		// MachineSet controller are racing to adopt Machines, see https://github.com/kubernetes-sigs/cluster-api/issues/7529
```